app/src/main/java/com/example/habitsmasher/ui/dashboard/AddHabitDialog.java

```diff
@@ -0,0 +1,186 @@
+                Date habitDate = null;
+                try {
+                    habitDate = inputDateFormatter.parse(habitDateInput);
+                } catch (ParseException e) {
+                    e.printStackTrace();
+                }
+                /**
+                 * Makes sure user does not enter blank fields,
+                 * and is capped on the number of characters for title and reason.
+                 */
+                HashMap<String, Habit> firebaseData = new HashMap<>();
+                Boolean properTitle = false;
+                Boolean properReason = false;
+                Boolean properDate = false;
+                if ((habitTitleInput.length()>0)&&(habitTitleInput.length()<=20)){
+                    properTitle = true;
+                }
+                if((habitReasonInput.length()>0)&&(habitReasonInput.length()<=30)){
+                    properReason = true;
+                }
+                if(!(habitDateInput.equals(""))){
+                    properDate = true;
+                }
+
+                /**
+                 * When user input meets the requirements for a proper habit title, habit reason, and habit date,
+                 * have the habit be added to the habit list back in dashboard fragment.
+                 *
+                 * Add the habit to our database.
+                 * Name of document in database will be the title of habit inputted by user.
+                 */
+
+                if ((properTitle==true)&&(properReason==true)&&(properDate==true)){
+                    _listener.addNewHabit(habitTitleInput, habitReasonInput, habitDate);
+                    firebaseData.put(habitTitleInput, new Habit(habitTitleInput, habitReasonInput, habitDate));
+                    collectionReference
+                            .document(habitTitleInput)
+                            .set(firebaseData)
+                            .addOnSuccessListener(new OnSuccessListener<Void>() {
+                                @Override
+                                public void onSuccess(Void unused) {
+                                    Log.d(TAG, "Data successfully added.");
+                                }
+                            })
+                            .addOnFailureListener(new OnFailureListener() {
+                                @Override
+                                public void onFailure(@NonNull Exception e) {
+                                    Log.d(TAG, "Data failed to be added." + e.toString());
+                                }
+                            });
+                    getDialog().dismiss();
+                } else{
+                    if (properTitle==false){
+                        habitTitleEditText.getText().clear();
+                        Toast.makeText(getActivity(), "Please enter a name for your habit (20 characters max).", Toast.LENGTH_SHORT).show();
```

Comment:
We can probably put a simpler error message here, something user friendly like: "_Habit name is too long_"